libbpf-tools: Add new feature doublefree

Bojun-Seo opened the pull request:
Add doublefree tool to detect double free. It could detect user level double
free error currently and can be expanded to detect kernel level double free
error. Followings are the usage and example.

Usage:
```
Usage: doublefree [OPTION...]
Detect and report doublefree error.

-c or -p is a mandatory option
EXAMPLES:
    doublefree -p 1234             # Detect doublefree on process id 1234
    doublefree -c a.out            # Detect doublefree on a.out
    doublefree -c 'a.out arg'      # Detect doublefree on a.out with argument

  -c, --command=COMMAND      Execute the command and detect doublefree
  -p, --pid=PID              Detect doublefree on the specified process
  -v, --verbose              Verbose debug output
  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.

Report bugs to https://github.com/iovisor/bcc/tree/master/libbpf-tools.

```
Example:
```
$ cat doublefree_generator.c
#include <unistd.h>
#include <stdlib.h>

int* foo() {
  return (int*)malloc(sizeof(int));
}

void bar(int* p) {
  free(p);
}

int main(int argc, char* argv[]) {
  sleep(10);
  int *val = foo();
  *val = 33;
  bar(val);
  *val = 84;
  bar(val);
  return 0;
}
$ gcc doublefree_generator.c
$ sudo ./doublefree -c a.out
2024-May-15 22:30:24 INFO Execute command: a.out(pid 99584)
Tracing doublefree... Hit Ctrl-C to stop
free(): double free detected in tcache 2

Allocation:
        #1 0x00564d1455819b foo+0x12 (/home/bojun/doublefree/libbpf-tools/a.out+0x119b)
        #2 0x00564d145581e3 main+0x27 (/home/bojun/doublefree/libbpf-tools/a.out+0x11e3)
        #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

First deallocation:
        #1 0x0070c4e94a53e0 free+0x0 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa53e0)
        #2 0x00564d145581fd main+0x41 (/home/bojun/doublefree/libbpf-tools/a.out+0x11fd)
        #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

Second deallocation:
        #1 0x0070c4e94a53e0 free+0x0 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa53e0)
        #2 0x00564d14558213 main+0x57 (/home/bojun/doublefree/libbpf-tools/a.out+0x1213)
        #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

```

Bojun-Seo committed:
libbpf-tools: Add doublefree tool

Add doublefree tool to detect double free. It could detect user level double
free error currently and can be expanded to detect kernel level double free
error. Followings are the usage and example.

Usage:

  $ ./doublefree -h
  Usage: doublefree [OPTION...]
  Detect and report doublefree error.

  -c or -p is a mandatory option
  EXAMPLES:
      doublefree -p 1234             # Detect doublefree on process id 1234
      doublefree -c a.out            # Detect doublefree on a.out
      doublefree -c 'a.out arg'      # Detect doublefree on a.out with argument

    -c, --command=COMMAND      Execute the command and detect doublefree
    -p, --pid=PID              Detect doublefree on the specified process
    -v, --verbose              Verbose debug output
    -?, --help                 Give this help list
        --usage                Give a short usage message
    -V, --version              Print program version

  Mandatory or optional arguments to long options are also mandatory or optional
  for any corresponding short options.

  Report bugs to https://github.com/iovisor/bcc/tree/master/libbpf-tools.

Example:

  $ cat doublefree_generator.c
  #include <unistd.h>
  #include <stdlib.h>
  #include <stdio.h>

  int* foo() {
    printf("foo\n");
    return (int*)malloc(sizeof(int));
  }

  void bar(int* p) {
    printf("bar\n");
    free(p);
  }

  int main(int argc, char* argv[]) {
    sleep(10);
    int *val = foo();
    *val = 33;
    bar(val);
    *val = 84;
    bar(val);
    return 0;
  }
  $ gcc doublefree_generator.c
  $ sudo ./doublefree -c a.out
  2024-May-15 22:30:24 INFO Execute command: a.out(pid 99584)
  Tracing doublefree... Hit Ctrl-C to stop
  free(): double free detected in tcache 2

  Allocation:
          #1 0x00564d1455819b foo+0x12 (/home/bojun/doublefree/libbpf-tools/a.out+0x119b)
          #2 0x00564d145581e3 main+0x27 (/home/bojun/doublefree/libbpf-tools/a.out+0x11e3)
          #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

  First deallocation:
          #1 0x0070c4e94a53e0 free+0x0 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa53e0)
          #2 0x00564d145581fd main+0x41 (/home/bojun/doublefree/libbpf-tools/a.out+0x11fd)
          #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

  Second deallocation:
          #1 0x0070c4e94a53e0 free+0x0 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa53e0)
          #2 0x00564d14558213 main+0x57 (/home/bojun/doublefree/libbpf-tools/a.out+0x1213)
          #3 0x0070c4e9429d90 [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

chenhengqi commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,158 @@
+}
+
+SEC("kretprobe/dummy_malloc")
+int BPF_KPROBE(malloc_return)
```
BPF_KRETPROBE

Bojun-Seo replied:
Thanks, I will change it from BPF_KPROBE into BPF_KRETPROBE.

chenhengqi commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,158 @@
+}
+
+SEC("kretprobe/dummy_calloc")
+int BPF_KPROBE(calloc_return)
```
Same here.

chenhengqi replied:
And below.

Bojun-Seo replied:
I'll change all of them

chenhengqi commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,158 @@
+	u64 addr = (u64)address;
+	struct doublefree_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+
+	if (info != NULL) {
```
```c
if (!info)
    return 0;
```

return earlier to avoid indention.

Bojun-Seo replied:
ok, I'll change as you said.

chenhengqi commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,594 @@
+/* default log level */
+static enum log_level log_level = ERROR;
+
+void sig_handler(int signo)
```
static ? Here and below.

Bojun-Seo replied:
There is a reason why I use static here and below. (below means static functions, right?)
I could remove static but I'd like to use static in here and below, if there is no problem.

I use this tool and other tools in my company, and manage the source code bit differently.
I compile every tool into one binary for user to use easily. So symbol can conflict on compilation if I don't use static.

Bojun-Seo replied:
I removed static on this enum value.

Bojun-Seo replied:
I have restored the static that I had previously removed. If there is a reason why static needs to be removed, please let me know.

eiffel-fl reviewed (commented):
Hi!

Thank you for it, this tool is cool!
I tested it and it worked fine:

```bash
$ ./a.out &
[2] 99365
$ sudo ./doublefree -p 99365
bar: 0x55c9bd2912a0
baz: 0x55c9bd2912a0
bazz: 84
free 84
free(): double free detected in tcache 2
[2]  + abort (core dumped)  ./a.out
Warn: Is this process alive? pid: 99365
Found double free...
Allocation happended on:
stack_id: 30033
        #1 0x0055c9bc24724b foo
        #2 0x0055c9bc2472a4 main
        #3 0x007fabf0d8c083 __libc_start_main


First deallocation happended on:
stack_id: 9246
        #1 0x007fabf0e026d0 free
        #2 0x0055c9bc2472be main
        #3 0x007fabf0d8c083 __libc_start_main


Second deallocation happended on:
stack_id: 3231
        #1 0x007fabf0e026d0 free
        #2 0x0055c9bc247236 baz
        #3 0x0055c9bc2472d4 main
        #4 0x007fabf0d8c083 __libc_start_main
```

I nonetheless have a question regarding the `memptrs` map.


Best regards.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,593 @@
+	syms = syms_cache__get_syms(syms_cache, env.pid);
+	lookup_key = -1;
+	while (!bpf_map_get_next_key(secondfd, &lookup_key, &next_key)) {
+		afd = bpf_map__fd(obj->maps.allocs);
+		firstfd = bpf_map__fd(obj->maps.first_deallocs);
```
Thanks for excellent review!
In fact, I put that two lines out of the loop before, but that caused some bugs(at that time).
But now I test it and find no bugs. So I'll apply your opinion.

Bojun-Seo commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,154 @@
+{
+	void *addr = NULL;
+	u64 pid = bpf_get_current_pid_tgid();
+	u64 *memptr64 = bpf_map_lookup_elem(&memptrs, &pid);
```
I should have to add some more code for `posix_memalign` entry.
I'll explain the code after I add it.

Bojun-Seo replied:
posix_memalign function does not return the pointer of allocated memory.
It gives the allocated pointer to user by using parameter passed.
https://man7.org/linux/man-pages/man3/posix_memalign.3.html
So this tool need saves the parameter address on `posix_memalign` entry.
(get memptr value and save it on `memptr` map)
And read the value pointed by the saved parameter on `posix_memalign` return
(read `memptr` map value and use it to read user memory)

As I mentioned above, I should add `posix_memalign` but I didn't
And now I add it.
Thanks again for meticulous review!

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,593 @@
+
+		file_name = strrchr(buf, '/') + 1;
+#ifdef WEBOS
+		if (sscanf(file_name, "libc-%f.so", &version) == 1) {
```
This issue is happening because of the difference of the path among platforms.
I think it would be better to add `libc` option like `gethostlatency` tool.

Bojun-Seo replied:
I remove the whole `libc_path` things rather use `get_pid_lib_path` written in `uprobe_helpers.c`.
It is nice portable function!

Bojun-Seo commented:
Add feature to detect doublefree on kernel(currently support arm and arm64)
Print report with symbol offset and module offset.

Bojun-Seo commented:
I found some issues on detecting kernel doublefree. So I remove it.
I'll implement kernel doublefree detection later.

Bojun-Seo commented:
Sounds great to merging tools. In fact, it could be possible to merge `memleak`, `lsan` and `doublefree`, since they hook(attach uprobes) the same points.
But my(and my company's) current interest is on embedded devices, and embedded devices are totally sensitive on memory usage.
So my team decided to split tools to use minimal memory.
But still, I think merging them is cool. So maybe I could merge them later.

Bojun-Seo commented:
I simplify the usage and the example, and I updated it on commit message as well as description.

eiffel-fl reviewed (commented):
Hi!


I have some comments but they are mainly nits.
I will test it again later.


Best regards.

Bojun-Seo commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,165 @@
+
+int gen_alloc_exit2(struct pt_regs *ctx, u64 address)
+{
+	struct doublefree_info_t info = {0, };
```
I thinks your suggestion seems better, I'll change it. Thanks!

Bojun-Seo commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,165 @@
+{
+	struct doublefree_info_t info = {0, };
+
+	if (address != 0) {
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces,
+			BPF_F_USER_STACK);
+		info.is_available = 1;
+		bpf_map_update_elem(&allocs, &address, &info, BPF_ANY);
+	}
+	return 0;
```
Thanks for the suggestion.
I'll change the code as you suggested.

Bojun-Seo commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,165 @@
+	} else if (info->is_available < 0) {
+		bpf_map_update_elem(&second_deallocs, &addr, &stack_id, BPF_ANY);
+	} else {
+		bpf_printk("This code should not be printed at any rate\n");
+		bpf_printk("Please check if something goes wrong\n");
```
I didn't know that `bpf_printk` could cause `dmesg` warning. Thanks for the information.

I agree with your opinion and I hope to get other's opinion as well before coding.
Fortunately(?) I found the similar code in existing `memleak` code.
So let's discuss this issue in already merged existing code.
I'll make a new issue and mention this comment and you.

Bojun-Seo replied:
On x86_64 machine Ubuntu 22.04.3 LTS,
I ran `doublefree` which contains `bpf_printk` but I cannot get warning message on `dmesg`.
And I ran `memleak` which contains `bpf_printk` as well(as I mentioned previously), but I still cannot get warning message on `dmesg`.
Could you please tell me the way to get warning message on `dmesg` when `bpf_printk` compiled eBPF code is loading?

Bojun-Seo replied:
Thank you for the checking!
`bpf_printk` is defined in file `bpftool/libbpf/src/bpf_helpers.h` and calls `bpf_trace_printk`.
So I guess we could add `bpf_printk_err` or `bpf_printk_dbg` to the file `bpf_helpers.h` and make it compiled only when `ERROR` or `DEBUG` macro is defined.
Since the file belongs to [libbpf](https://github.com/libbpf/libbpf), I'll make issue on this project.
Or you can make an issue(it's your idea).

Bojun-Seo replied:
I changed this program to use `bpf_perf_event`. And I carry error code on event instead of print it inside bpf code. So this issue is resolved.

Bojun-Seo commented on libbpf-tools/doublefree.bpf.c:
```diff
@@ -0,0 +1,165 @@
+SEC("kprobe/dummy_posix_memalign")
+int BPF_KPROBE(posix_memalign_entry, void **memptr, size_t alignment, size_t size)
+{
+	u64 memptr64 = (u64)(size_t)memptr;
```
I'll change it to use one casting, thanks!

Bojun-Seo replied:
I found the similar double casting code on file `bcc/tools/memleak.py`
```c
307 int posix_memalign_enter(struct pt_regs *ctx, void **memptr, size_t alignment,
308                          size_t size) {
309         u64 memptr64 = (u64)(size_t)memptr;
310         u64 pid = bpf_get_current_pid_tgid();
311
312         memptrs.update(&pid, &memptr64);
313         return gen_alloc_enter(ctx, size);
314 }
```
I guess this double casting is to support 32-bits architectures.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,556 @@
+	}
+
+	/* malloc */
+	func_off = get_elf_func_offset(libc_path, "malloc");
+	if (func_off < 0) {
+		p_err("Failed to get func offset");
+		return func_off;
+	}
+	obj->links.malloc_return =
+			bpf_program__attach_uprobe(obj->progs.malloc_return,
+			true, env.pid ? : -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.malloc_return);
+	if (err) {
+		p_err("Failed to attach malloc_return");
+		return err;
+	}
```
I totally agree with you. I'll factorize it. Thanks

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,556 @@
+		}
+	} while (check_doublefree(syms_cache) == 0);
+
+	/* cleanup */
```
It is `main` function. So I add `return` or `exit` for each error cases instead of set label and add `goto` statement.

Bojun-Seo replied:
> Sadly, if you return you may leak some resources.

As far as I know, linux OS close every file belongs to the process in process exit. And the file I mentioned means abstraction of every resources. So, this coding pattern(not cleanup and return or exit in `main` function) does not leak resources, even though `lsan`, `asan` as well as `valgrind` reports it as leak.
Do you think that there might be a resource leak contrary to what I said?

Bojun-Seo replied:
Great! And thank you for checking eBPF resource deallocation is fine with process exit.

eiffel-fl reviewed (commented):
Hi!


I have one comment regarding the macros which I think can be replaced by a function.
Otherwise, it looks good to me.
Let me know about my comment and I will test it later.


Best regards.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+		_CHECK_PROGRAM(obj, func_name##_entry); \
+	} while (false)
+
+#define ATTACH_URETPROBE_CHECKED(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_CHECK_OFFSET(func_off); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_return, true, func_off, pid); \
+		_CHECK_PROGRAM(obj, func_name##_return); \
+	} while (false)
```
Sounds great! I always appreciate your valuable opinion.
I prefer the second code you suggested(using struct).
And I think it would be better to pass `struct probe` instance(using pointer) instead of each members to `attach_uprobe` function, like followings.
```C
for (int i = 0; i < sizeof(probes) / sizeof(probes[0]); i++) {
	attach_uprobe(&probes[i], libc_path);
}
```
What you think?

Bojun-Seo replied:
I changed uprobe attaching codes. I tried implementing it while incorporating your opinion, but in a slightly different way.

ekyooo reviewed (commented):
"ERROR Failed to poll perf_buffer" in usage example always displayed?

ekyooo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+	int deallocs_fd = bpf_map__fd(obj->maps.deallocs);
+
+	if (e->err == -1) {
+		p_err("This message should not be printed..");
```
How about removing code that shouldn't happen?

Bojun-Seo replied:
It's a good point. However, thinking that something will never be executed is just the developer's assumption. In software development, it is not uncommon for code that should not be executed to actually be executed. Especially when contributing code to open source projects, where multiple developers are involved, there is a higher possibility of unintentionally executing code that should not be executed. Therefore, I believe that this code should be kept as a precaution for future mistakes by someone else. However, the message itself seems lengthy and lacks specific details. I will consider possible improvements.

Bojun-Seo replied:
I didn't remove the message but changed it.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	p_err("handle_lost_events");
```
Thanks for your review. I'll change the code like followings, referring `opensnoop.c`
```C
static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
{
        p_err("Lost %llu events on CPU #%d!", lost_cnt, cpu);
}
```

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+				env.perf_max_stack_depth * sizeof(__u64));
+	bpf_map__set_max_entries(obj->maps.stack_traces,
+				 env.stack_storage_size);
+	err = doublefree_bpf__load(obj);
```
I'll add line breaks not only here, but also in other places.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+	if (cmd == NULL) {
+		p_warn("Invalid argument: command not found");
+		exit(-1);
+	} else
```
I'll add `{}` in `else` case.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+
+	err = bpf_map_lookup_elem(fd, &key, &val);
+	if (err < 0) {
+		p_err("Failed to lookup elem on fd");
```
You are right. I'll change the code. Thanks!

Bojun-Seo replied:
I'd rather change the code to print `strerror(errno)` instead of `fd`

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+		ptr = strtok(NULL, delim);
+	} else {
+		p_err("failed to exec %s", cmd);
+		goto cleanup;
```
I can and will remove duplicates but cannot handle the error cases first.
Because error case can be checked after tokenizing or calling `fork`.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+	pid_t pid;
+	int stack_storage_size;
+	int perf_max_stack_depth;
+	/* unit: second */
```
I will remove the comment.
Initially, the comment was about the `interval` value, which was removed during the transition to an event-driven approach. Consequently, the corresponding comment should have been deleted, but it was inadvertently left behind.
Thanks for the catch!

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+		return;
+	}
+	syms = syms_cache__get_syms(syms_cache, env.pid);
+	if (syms != NULL) {
```
Sounds nice, that change could reduce one indent as well.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+static volatile sig_atomic_t exiting = 0;
+
+static struct env {
+	/* main process's tid */
```
Seems too excessive explanation. I'd rather remove the comment. Thanks!

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+		return pid;
+	} else {
+		p_err("failed to exec %s", cmd);
+		exit(-1);
```
I agree with you. I'll change it to `return` rather than `exit`.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,473 @@
+			p_err("Failed to poll perf_buffer");
+			break;
+		}
+		if (getpgid(env.pid) < 0) {
```
There is no need to limit it to the `ESRCH` errno case according to the following manual page.
https://linux.die.net/man/2/getpgid
`getpgid` returns `-1` on error and errno is set appropriately.
And there is only one errno for `getpgid`, which is `ESRCH`.
Which means, if `getpgid` fails, the errno should be `ESRCH`.

Thanks for your question.

Bojun-Seo replied:
And I found that calling `getpgid` is not necessary. So I remove that code.

Bojun-Seo reviewed (commented):
> "ERROR Failed to poll perf_buffer" in usage example always displayed?

Yes, but it would be better not to be printed in normal case.

eiffel-fl reviewed (approved):
Hi!

I just found some small nits, but I think we are ready to go:

```bash
$ sudo ./doublefree -c /tmp/doublefree_generator                                                                                                                       (remotes/bojun/doublefree) *%
[sudo] Mot de passe de francis : 
Désolé, essayez de nouveau.
[sudo] Mot de passe de francis : 
2024-Feb-29 16:20:13 INFO Execute command: /tmp/doublefree_generator(pid 94846)
Tracing doublefree... Hit Ctrl-C to stop
free(): double free detected in tcache 2

Allocation:
        #1 0x0056477a65b19b foo+0x12
        #2 0x0056477a65b1e3 main+0x27
        #3 0x007f2259229d90 [unknown]

First deallocation:
        #1 0x007f22592a53e0 free+0
        #2 0x0056477a65b1fd main+0x41
        #3 0x007f2259229d90 [unknown]

Second deallocation:
        #1 0x007f22592a53e0 free+0
        #2 0x0056477a65b213 main+0x57
        #3 0x007f2259229d90 [unknown]
```

Best regards.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,461 @@
+		free(argv);
+
+		return -1;
+	} else {
+		/* Failed to create child process */
+		return -1;
+	}
```
Sounds nice. I'll apply it.

eiffel-fl reviewed (commented):
Hi!

I just have one comment with your latest push, but you can ignore it.

Best regards.

Bojun-Seo commented on libbpf-tools/doublefree.c:
```diff
@@ -0,0 +1,467 @@
+	for (i = 0; i < env.perf_max_stack_depth && ip[i]; ++i) {
+		printf("\t#%zu %#016lx", i + 1, ip[i]);
+		err = syms__map_addr_dso(syms, ip[i], &sinfo);
+		if (!err) {
+			if (sinfo.sym_name)
+				printf(" %s+0x%lx", sinfo.sym_name, sinfo.sym_offset);
+			else
+				printf(" [unknown]");
+
+			printf(" (%s+0x%lx)\n", sinfo.dso_name, sinfo.dso_offset);
+		}
```
It seems more readable. I'll change the code as you suggested.
Thank you!

Bojun-Seo replied:
I found that the location of `newline` is not proper in my code.
So I change it. I can find it thanks for your help!

baonq-me commented:
> I just rebased this PR and I have confirmed that this feature is still working well, and I hope it gets merged into the mainline.
> 
> I will discuss this PR at OSS Korea 2025. https://osskorea2025.sched.com/event/2913v/detecting-double-free-with-bpf-bojun-seo-lg-electronics?iframe=yes&w=100%&sidebar=yes&bg=no If you're interested, I would love for you to join and share your thoughts.

Interesting. I come here from the OSS Korea event.

Bojun-Seo commented:
This patch is an updated version of the original submission. The changes listed below are assisted by GitHub Copilot, and the full patch is reviewed by Copilot before posting.

Changes compared to the original commit:

Bug fixes:
- Wrong BPF probe macros: The original code used BPF_KPROBE/BPF_KRETPROBE for user-space probes, which are incorrect. Changed to BPF_UPROBE/BPF_URETPROBE.
- calloc argument order: strlen(cmd) / 2 + 1 was off by one — the do/while loop always appends a NULL terminator, which could write one element past the allocated array. Changed to strlen(cmd) / 2 + 2 with an explanatory comment.
- get_stackid type mismatch: The single get_stackid() function was used for both allocs (value: struct doublefree_info_t, 8 bytes) and frees (value: u32, 4 bytes) maps. This only worked accidentally due to little-endian layout. Split into get_alloc_stackid() and get_free_stackid() so each uses the correct value type.

Improvements:
- Runtime timestamps: Replaced compile-time __DATE__/__TIME__ macros with time()/strftime() so log timestamps reflect when the tool actually runs, not when it was compiled.
- ARRAY_SIZE macro: Used ARRAY_SIZE(probes) instead of a hardcoded count for the probe attachment loop.
Naming: Renamed dealloc/deallocs/deallocation to free/frees/free throughout for consistency with standard C terminology.

Changes:
- Use a non-atomic counter for simplicity. This could cause a race condition, but it is very unlikely. It can be improved in the future.
